pep-0xxx.rst

```diff
@@ -0,0 +1,727 @@
+
+The ``Context`` abstract base class has the following class definition::
+
+    TLSBufferObject = Union[TLSWrappedSocket, TLSWrappedBuffer]
+    ServerNameCallback = Callable[[TLSBufferObject, Optional[str], Context], Any]
+
+    class _BaseContext(metaclass=ABCMeta):
+
+        @property
+        @abstractmethod
+        def validate_certificates(self) -> bool:
+            """
+            Whether to validate the TLS certificates. This switch operates at a
+            very broad scope: either validation is enabled, in which case all
+            forms of validation are performed including hostname validation if
+            possible, or validation is disabled, in which case no validation is
+            performed.
+
+            Not all backends support having their certificate validation
+            disabled. If a backend does not support having their certificate
+            validation disabled, attempting to set this property to ``False``
+            will throw a ``TLSError``.
+            """
+
+        @validate_certificates.setter
+        @abstractmethod
+        def validate_certificates(self, val: bool) -> None:
+          pass
+
+        @abstractmethod
+        def register_certificates(self,
+                                  certificates: str,
+                                  key=None: Optional[str],
+                                  password=None: Optional[Callable[[], Union[AnyStr, bytearray]]]) -> None:
+            """
+            Loads a certificate, a number of intermediate certificates, and the
+            corresponding private key. These certificates will be offered to
+            the remote peer during the handshake if required.
+
+            The ``certificates`` argument must be a bytestring containing the
+            PEM-encoded certificates. The first PEM-encoded certificate must be
+            the leaf certificate. All subsequence certificates will be offered
+            as intermediate additional certificates.
+
+            The ``key`` argument, if present, must contain the PEM-encoded
+            private key associated with the leaf certificate. If not present,
+            the private key will be extracted from ``certificates``.
+
+            The ``password`` argument may be a function to call to get the
+            password for decrypting the private key. It will only be called if
+            the private key is encrypted and a password is necessary. It will
+            be called with no arguments, and it should return a string, bytes,
+            or bytearray. If the return value is a string it will be encoded as
+            UTF-8 before using it to decrypt the key. Alternatively a string,
+            bytes, or bytearray value may be supplied directly as the password
+            argument. It will be ignored if the private key is not encrypted
+            and no password is needed.
+            """
+
+        @abstractmethod
+        def set_ciphers(self, ciphers: List[Ciphers]) -> None:
```

Comment:
I have some thoughts written down on a possible way to support some kind of cipher filtering in a similar project: https://github.com/sfackler/rust-native-tls/issues/4#issue-173046053.

The usefulness does seem to be somewhat limited compared to full cipher selection, but it might still be worth it to be able to do things like disable non-ephemeral key exchange, for example.